### PR TITLE
 BHBC-2300: Apply 'Viewer' and 'Editor' Permissions to Project / Survey Views

### PR DESCRIPTION
**Pull request by curtisupshall**
# Overview

As a Project Lead, I want to make sure that people I’ve invited to my team can only see the functions their role allows them to so they aren’t getting confused as to why some functionality appears broken.

## Links to Jira tickets

 - https://quartech.atlassian.net/browse/BHBC-2300

## Description of relevant changes

 - Created an API method to get the user's current roles for the given project
 - Created a Project Auth State context to determine the auth state for project and survey views
 - Created a component guard to hide UI elements based on accepted project roles
 - Created a route guard to protect app routes based on accepted project roles. Current protected routes include:
   - Project Manage Users page
   - Edit Project page
   - Create Survey page
   - Edit Survey page

## Testing Procedures

0. Create a BCeID Basic in the BCeID Dev environment. You can do this here: https://www.development.bceid.ca/register/.
1. From your IDIR account, add your BCeID user to a project as a **‘Viewer’.**
2. Go to Manage Users. Ensure that your newly added BCeID user does not have admin permissions by changing their role to _Creator._
3. Open a new Incognito (Chrome)/Private Browsing (Firefox)/InPrivate (Edge) window **(Ctrl + Shift + N)**. Sign into SIMS as your BCeID user.
4. View the project as the 'Viewer' user. It should:

 - [x] NOT show the following interactive components:
    - [x] Add Survey
    - [x] Upload Documents
    - [x] Delete Document (see individual document row items)
    - [x] Edit Location
  - [x] NOT be able to click **Project Settings**

5. From the project view, change your browser's URL from `/details` to `/edit`. It should:

 - [x] Show the Forbidden page.

6. From the project view, change your browser's URL from `/details` to `/users`. It should:

 - [x] Show the Forbidden page.

7. Navigate to a survey on the same project. It should:

 - [x] NOT show the following interactive components:
    - [x] Import Observations
    - [x] Delete Observations
    - [x] Import Summary Results
    - [x] Delete Summary Results
    - [x] Upload Documents
    - [x] Delete Document (see individual document row items)
    - [x] Edit Study Area
  - [x] NOT allow you to click **Survey Settings**

8. From the survey view, change your browser's URL from `/details` to `/edit`. It should:

 - [x] Show the Forbidden page.

9. From the survey view, change your browser's URL from `/{surveyId}/details` to `/create`. It should:

 - [x] Show the Forbidden page.

10. Change the project role of your BCeID user to be an **‘Editor’.**
11. View the project as the 'Editor' user. It should:

 - [x] NOT show the following project settings options:
    - Delete Project
    - Manage Project Team

12. From the project view, change your browser's URL from `/details` to `/users`. It should:

 - [x] Show the Forbidden page.

### CODE REVIEW

**Comment by NickPhura** [comment]
@curtisupshall Just skimmed what you have here, looks pretty good so far. A few thoughts, semi jumbled.

It might be ideal to turn that participant guard context into something more generic, like `projectAuthStateContext` (mirroring the existing `authStateContext`). It could then wrap the project routes, similar to the project context. Then we can leverage it however we want in the children (could even look into project route guards which we don't have atm). We could also look into a specific endpoint for fetching the project user info for that specific `projectId`, in case the user is on a ton of projects.

(I think the only funny project route atm is the edit one, as it doesn't fall under the `:id` path, but probably should. And then if we do build a project route guard, it could be protected the same way).

Overall I think its similar to the work you did with the project/survey contexts really and fixing the routers a bit in the process, etc.

**Comment by curtisupshall** [comment]
@NickPhura 
> It might be ideal to turn that participant guard context into something more generic, like `projectAuthStateContext` (mirroring the existing `authStateContext`). It could then wrap the project routes, similar to the project context.

~~Sounds like a good approach!~~

Hmm, now I wonder if `projectParticipantStateContext` would be a better candidate... 🤔 Beyond checking the participant record, is there any metric not covered by `keycloakWrapper` by which we could authenticate for a project? Conversely, a project participant context could be used for authentication in addition to other project role functionallity

> We could also look into a specific endpoint for fetching the project user info for that specific `projectId`, in case the user is on a ton of projects.

This was my initial thinking; I wrote the `project/{projectId}/participants/self.ts` endpoint to get the participant info for the user.

> Overall I think its similar to the work you did with the project/survey contexts really and fixing the routers a bit in the process, etc.

Yup, agreed.

**Comment by NickPhura** [comment]
> @NickPhura
> 
> > It might be ideal to turn that participant guard context into something more generic, like `projectAuthStateContext` (mirroring the existing `authStateContext`). It could then wrap the project routes, similar to the project context.
> 
> ~Sounds like a good approach!~
> 
> Hmm, now I wonder if `projectParticipantStateContext` would be a better candidate... 🤔 Beyond checking the participant record, is there any metric not covered by `keycloakWrapper` by which we could authenticate for a project? Conversely, a project participant context could be used for authentication in addition to other project role functionallity
> 
> > We could also look into a specific endpoint for fetching the project user info for that specific `projectId`, in case the user is on a ton of projects.
> 
> This was my initial thinking; I wrote the `project/{projectId}/participants/self.ts` endpoint to get the participant info for the user.
> 
> > Overall I think its similar to the work you did with the project/survey contexts really and fixing the routers a bit in the process, etc.
> 
> Yup, agreed.

Lets chat in Discord. I suspect we are on the same track tbh and its mostly just a semantics thing.

**Comment by AlfredRosenthal** [comment]
Might be out of scope but running your branch I see this when I try to manage users: 
Screenshot 2023-04-13 124926

**Comment by curtisupshall** [comment]
> Might be out of scope but running your branch I see this when I try to manage users: Screenshot 2023-04-13 124926

Fixed by https://github.com/bcgov/biohubbc/pull/999.

File: api/src/paths/project/{projectId}/participants/self.test.ts
Comment (curtisupshall) [inline api/src/paths/project/{projectId}/participants/self.test.ts:1]: Just need to address these tests; they're copied from another file.

**Review by NickPhura — DISMISSED** [review]

**Review by NickPhura — DISMISSED** [review]

File: app/src/features/projects/view/ProjectHeader.tsx [hunk 1]
```diff
@@ -234,7 +226,11 @@ const ProjectHeader = () => {
                 aria-controls="projectSettingsMenu"
                 aria-haspopup="true"
                 style={{ marginLeft: '0.5rem' }}
-                onClick={handleClick}>
+                onClick={handleClick}
+                disabled={
+                  !hasProjectRole([PROJECT_ROLE.PROJECT_EDITOR, PROJECT_ROLE.PROJECT_LEAD]) &&
```
Comment (KjartanE) [inline app/src/features/projects/view/ProjectHeader.tsx]: image
This button should be hidden from the viewer not disabled

File: app/src/features/surveys/view/SurveyHeader.tsx [hunk 2]
```diff
@@ -206,7 +210,11 @@ const SurveyHeader = () => {
                   startIcon={<Icon path={mdiCogOutline} size={1} />}
                   endIcon={<Icon path={mdiChevronDown} size={1} />}
                   onClick={openSurveyMenu}
-                  style={{ marginLeft: '0.5rem' }}>
+                  style={{ marginLeft: '0.5rem' }}
+                  disabled={
```
Comment (KjartanE) [inline app/src/features/surveys/view/SurveyHeader.tsx]: Same here, Button should be hidden not disabled

**Review by KjartanE — COMMENTED** [review]
image

\app\src\features\surveys\view\survey-observations\components\NoObservationsCard.tsx
\app\src\features\surveys\view\summary-results\components\NoSummaryResults.tsx

THese import button are still available

File: app/src/features/surveys/SurveyRouter.tsx [hunk 3]
```diff
@@ -12,21 +12,17 @@ import AppRoute from 'utils/AppRoute';
...
-      <AppRoute exact path="/admin/projects/:id/surveys/:survey_id/observations" layout={ProjectsLayout}>
-        <SurveyPage />
-      </AppRoute>
-
-      <AppRoute exact path="/admin/projects/:id/surveys/:survey_id/results" layout={ProjectsLayout}>
-        <SurveyPage />
-      </AppRoute>
+      <Route exact path="/admin/projects/:id/surveys/:survey_id/edit">
+        <ProjectsLayout>
+          <EditSurveyPage />
```
Comment (KjartanE) [inline app/src/features/surveys/SurveyRouter.tsx]: image
Edit Survey page still accessible by Viewer

**Review by KjartanE — APPROVED** [review]
Functionally works well, no issues noticed. 🐈

**Review by NickPhura — APPROVED** [review]

**Pull request merged**